app/src/main/java/org/mozilla/focus/browser/FirefoxProgressBar.kt

```diff
@@ -26,23 +25,34 @@ private const val HIDE_ANIMATION_DURATION_MILLIS = 250L
 private val HIDE_AFTER_MILLIS = TimeUnit.SECONDS.toMillis(3)
 
 class FirefoxProgressBar @JvmOverloads constructor(
-        context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
-    : LinearLayout(context, attrs, defStyle) {
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : LinearLayout(context, attrs, defStyle), Session.Observer {
 
     private val hideHandler = FirefoxProgressBarHideHandler(this)
+    private var session: Session? = null
 
     fun initialize(browserFrag: BrowserFragment) {
-        browserFrag.session.url.observe(browserFrag, Observer { sessionUrl -> url.text = sessionUrl })
-        browserFrag.session.loading.observe(browserFrag, object : NonNullObserver<Boolean>() {
-            public override fun onValueChanged(value: Boolean) {
-                val isLoading = value
-                if (isLoading) {
-                    showBar()
-                } else {
-                    scheduleHideBar()
-                }
+        session = browserFrag.session.also {
+            it.register(observer = this, owner = browserFrag)
+        }
+
+        browserFrag.session.register(observer = this, owner = browserFrag)
+    }
+
+    override fun onLoadingStateChanged() {
+        session?.let {
```

Comment:
nit: In this case, to avoid indentation, I prefer:
```kotlin
val session = session ?: return
if (session.isLoading) { ...
```

Reply:
With the `lateinit` change session is not nullable anymore. I'l update that.